blog/_src/posts/2017-05-12-artifacts-for-semantics.md

```diff
@@ -0,0 +1,155 @@
+   much effort for readers.
+3. Put it online! Don't require installing any software! Conveniently,
+   implementing 2 is also made easier once done online. By using OCaml, we were
+   able to easily use the
+   excellent [js_of_ocaml](http://ocsigen.org/js_of_ocaml/), though there are
+   other ways of getting Javascript versions of your artifacts (without actually
+   having to write them in Javascript). 
+
+The first was done by Gabriel, who wrote a grammar
+using [Menhir](http://gallium.inria.fr/~fpottier/menhir/), and then equipped it
+with custom parsing error messages that provide much better feedback when there
+are typos in what people are trying. We also wrote a pretty-printer using
+the [PPrint](http://gallium.inria.fr/blog/first-release-of-pprint/) library, so
+we could show intermediate program states through the UI. After writing this, we
+were able to convert our existing suite of test cases and examples to be
+written textually, which was a huge improvement for us as well! These and other
+tests were used to ensure that the parser/pretty-printer would round-trip properly.
+
+For the interface, I built a simple web page that had
+the [CodeMirror](https://codemirror.net/) editor equipped with a very simple
+syntax highlighter (8 lines of code to highlight keywords & atoms, plus a
+CodeMirror extension to highlight matching brackets) and error highlighting
+(which is triggered by the OCaml code). I then made a simple "machine state" UI
+that showed, in pretty-printed format, the heap, stack, registers, context, and
+redex. On the OCaml side, when the "run" button is clicked, we parse and
+typecheck and, assuming no errors occur, store the current state as our "history".
+As the user clicks forward or backwards, we run the step function and append to the history of
+states or pop states off of the history. In total, there
+are
+[50 lines of Javascript](https://github.com/dbp/funtal/blob/032be70f33f77e80f4fab7e62016bfabf96476f3/artifact/index.html#L246) and
+about
+[150 lines of OCaml](https://github.com/dbp/funtal/blob/032be70f33f77e80f4fab7e62016bfabf96476f3/web.ml) that
+handle the logic of the editor.
+
+Putting it online was very easy, based on the choice of tools used earlier. We
+compile the main file
+([web.ml](https://github.com/dbp/funtal/blob/032be70f33f77e80f4fab7e62016bfabf96476f3/web.ml))
+to Javascript using [js_of_ocaml](http://ocsigen.org/js_of_ocaml/), and it pulls in the parser, type-checker,
+interpreter, examples, etc. The rest of the artifact is a single html file, a
+CSS file, and a few javascript files for CodeMirror. It requires no server
+backend, is easy to archive and save, and will even run on smartphones! 
+
+The total time spent implementing the artifact was a small fraction of the time
+spent on the paper, and while it was not in any critical way essential for the
+success of the paper, it does make the paper much easier to read, and we would
```

Comment:
Did you find minor mistakes/typos in the paper while implementing the operational semantics or type checker? That could be useful to point out.